lms/djangoapps/courseware/access.py

```diff
@@ -273,6 +274,12 @@ def _can_enroll_courselike(user, courselike):
     if _has_staff_access_to_descriptor(user, courselike, course_key):
         return ACCESS_GRANTED
 
+    # Access denied when the user cannot self enroll (ccx courses). This feels very similar to being "invitation only"
+    # but some code treats these differently, so I'm not unifying the logic yet.
+    if not can_self_enroll_in_course(course_key):
+        debug("Deny: cannot self enroll")
+        return ACCESS_DENIED
+
```

Comment:
This is adding a new check into `has_access()` which is used in a lot of places. But it seems like a sensible check. It was historically not super necessary since the course about page just redirects to the dashboard if a course is not self enrollable, so we never got as far as asking whether the user can enroll and checking `has_access`.

The MFE doesn't specifically check for this case and redirect though. So we may as well not show an enroll button if this case does apply. And logically, it makes sense that if you can't self enroll, the `can_enroll` access is denied.

This method is simple and looks like:
```
def can_self_enroll_in_course(course_key):
    """
    Returns True if the user can enroll themselves in a course.

    Note: an example of a course that a user cannot enroll in directly
    is a CCX course. For such courses, a user can only be enrolled by
    a CCX coach.
    """
    if hasattr(course_key, 'ccx'):
        return False
    return True
```

So, :shrug: